include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -431,68 +431,79 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 
-//-----------------------------------------------------------------------
-// future and helper classes for async pattern/algorithm
-//-----------------------------------------------------------------------
+template <typename _Event>
+void __wait_event(_Event);
 
-// TODO: towards higher abstraction and generic future. implementation specific sycl::event should be hidden
-struct __future_base
+void
+__wait_event(sycl::event __e)
 {
-    sycl::event __my_event;
-
-    __future_base() = default;
-    __future_base(sycl::event __e) : __my_event(__e) {}
-    void
-    wait()
-    {
 #if !ONEDPL_ALLOW_DEFERRED_WAITING
-        __my_event.wait_and_throw();
+    __e.wait_and_throw();
 #endif
-    }
-    operator sycl::event() const { return __my_event; }
-};
+}
 
 template <typename _T>
-class __future : public __future_base
+constexpr auto
+__get_value(sycl::buffer<_T>& __buf)
 {
-    ::std::size_t __result_idx;
-    sycl::buffer<_T> __data;
-
-  public:
-    __future(sycl::event __e, size_t __o, sycl::buffer<_T> __b)
-        : __par_backend_hetero::__future_base(__e), __result_idx(__o), __data(__b)
-    {
-    }
+    //according to a contract, returned value is one-element sycl::buffer
+    return __buf.template get_access<access_mode::read>()[0];
+}
 
-    _T
-    get()
-    {
-        return __data.template get_access<access_mode::read>()[__result_idx];
-    }
-    template <class _Tp, class _Enable>
-    friend class oneapi::dpl::__internal::__future;
-};
+template <typename _T>
+constexpr auto
+__get_value(_T& __val)
+{
+    return __val;
+}
 
-template <>
-class __future<void> : public __future_base
+//A contract for future class: <sycl::event or other event, a value or sycl::buffers...>
+//Impl details: inheretance (private) instead of aggregation for enabling the empty base optimization.
+template <typename _Event, typename... _Args>
+class __future : private std::tuple<_Args...>
 {
-    ::std::unique_ptr<oneapi::dpl::__internal::__lifetime_keeper_base> __tmps;
+    _Event __my_event;
+
+    template <typename... _Prms>
+    struct __value_ckeck
+    {
+        static constexpr bool __is_value = sizeof...(_Prms) > 0;
+    };
 
   public:
-    template <typename... _Ts>
-    __future(sycl::event __e, _Ts... __t) : __future_base(__e)
+    __future(_Event __e, _Args... __args) : std::tuple<_Args...>(__args...), __my_event(__e) {}
+    __future(_Event __e, std::tuple<_Args...> __t) : std::tuple<_Args...>(__t), __my_event(__e) {}
+
+    auto
+    event() const
     {
-        if (sizeof...(__t) != 0)
-            __tmps = ::std::unique_ptr<oneapi::dpl::__internal::__lifetime_keeper<_Ts...>>(
-                new oneapi::dpl::__internal::__lifetime_keeper<_Ts...>(__t...));
+        return __my_event;
     }
+    operator _Event() const { return event(); }
     void
+    wait()
+    {
+        __wait_event(event());
+    }
+
+    template <typename _T = __value_ckeck<_Args...>, ::std::enable_if_t<_T::__is_value, int> = 0>
```

Comment:
should be `__value_check` instead of `__value_ckeck`. By the way, if this part of code is needed to be rewritten anyway, should we use `::std::tuple_size` instead?

Reply:
yes, we may use `::std::tuple_size_v` instead.